Modify memory option rounding to avoid overflow

chrisdennis opened the pull request:
~Add GET_RAW_MEM_VALUE and associated setRawMemoryOptionToOptElse helper function. Adopt for the `-XX:MaxDirectMemorySize` parsing to avoid rounding values passed to the Java layer beyond the internal signed 64-bit limit.~

Memory option values are rounded up to be pointer-width aligned unless this would lead to a value greater than U_32_MAX or I_64_MAX on 32 and 64 bit platforms respectively. In these cases we truncate (round down) to the next pointer-width aligned value.
    
At the same time we add a bounds check on the MaxDirectMemorySize value on 64 bit platforms to ensure it maps safely to a Java long before passing in to the JVM initialization.

Issue https://github.com/eclipse-openj9/openj9/issues/22480

JasonFengJ9 commented:
@chrisdennis thanks your contribution.

This PR addressed the issue for `-XX:MaxDirectMemorySize=9223372036854775807`, however there is still a difference between HotSpot VM and OpenJ9 for `-XX:MaxDirectMemorySize=9223372036854775808` (an existing behaviour).
 
HotSpot JDK25 VM validates the system property value before its usage.
```
java -XX:MaxDirectMemorySize=9223372036854775808 -version
uint64_t MaxDirectMemorySize=9223372036854775808 is outside the allowed range [ 0 ... 9223372036854775807 ]
Improperly specified VM option 'MaxDirectMemorySize=9223372036854775808'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

OpenJ9 allows the invalid `9223372036854775808` value to be saved into `vm->directByteBufferMemoryMax`, and the `NumberFormatException` is thrown later by `Long.parseLong()`.  
```
java --add-exports=java.base/jdk.internal.misc=ALL-UNNAMED -XX:MaxDirectMemorySize=9223372036854775808 -jar TestMiscVM.jar
Exception in thread "(unnamed thread)" java/lang/NumberFormatException: For input string: "9223372036854775808"
	at java/lang/NumberFormatException.forInputString (java.base@25/NumberFormatException.java:67)
	at java/lang/Long.parseLong (java.base@25/Long.java:594)
	at java/lang/Long.parseLong (java.base@25/Long.java:698)
	at jdk/internal/misc/VM.saveProperties (java.base@25/VM.java:258)
	at java/lang/System.ensureProperties (java.base@25/System.java:423)
	at java/lang/System.afterClinitInitialization (java.base@25/System.java:227)
	at java/lang/Thread.initialize (java.base@25/Thread.java:2742)
	at java/lang/Thread.<init> (java.base@25/Thread.java:2806)
```

I think OpenJ9 should match HotSpot and validate the system property value before its usage.

pshipton commented on runtime/vm/jvminit.c:
```diff
@@ -2557,6 +2557,11 @@ VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved)
 				parseErrorOption = VMOPT_XXMAXDIRECTMEMORYSIZEEQUALS;
 				goto _memParseError;
 			}
+			if ((vm->directByteBufferMemoryMax != ~(UDATA)0) && (vm->directByteBufferMemoryMax > ((1 << 63) - 1))) {
```
I don't think this works for 32-bit.

chrisdennis replied:
You're right... I'm in the process of reworking this to be more platform/width independent.

pshipton commented on runtime/util/vmargs.c:
```diff
@@ -626,6 +625,12 @@ optionValueOperations(J9PortLibrary *portLibrary, J9VMInitArgs* j9vm_args, IDATA
 							return OPTION_MALFORMED;
 					}
+
+					if (action==GET_MEM_VALUE) {
```
Pls add spaces around `==`.

chrisdennis replied:
I assume both of these hold only for the lines I'm touching? I originally went with this formatting to maintain consistency with the existing code.

pshipton commented on runtime/util/vmargs.c:
```diff
@@ -626,6 +625,12 @@ optionValueOperations(J9PortLibrary *portLibrary, J9VMInitArgs* j9vm_args, IDATA
+
+					if (action==GET_MEM_VALUE) {
+						oldValue = value;
+						value = (value +  sizeof(UDATA) - 1) & ~(sizeof(UDATA) - 1);		/* round to nearest pointer value */
```
Pls remove the extra space after `+`.

pshipton commented on runtime/util/vmargs.c:
```diff
@@ -626,6 +625,12 @@ optionValueOperations(J9PortLibrary *portLibrary, J9VMInitArgs* j9vm_args, IDATA
+					if (action==GET_MEM_VALUE) {
+						oldValue = value;
+						value = (value +  sizeof(UDATA) - 1) & ~(sizeof(UDATA) - 1);		/* round to nearest pointer value */
+						if (value < oldValue) return OPTION_OVERFLOW;
```
I don't believe the coding standards allow this. Pls add `{}`.
https://github.com/eclipse-omr/omr/blob/master/doc/CodingStandard.md

chrisdennis commented:
@pshipton I want to do something like this, but obviously this isn't using the typedefs from OMR-land. There are defines for UDATA_MAX and I_64_MAX but they're not suitable for use in a preprocessor condition.  Any recommendations on the best approach here?
```c
#if UINTPTR_MAX > INT64_MAX
			if ((vm->directByteBufferMemoryMax != ~(UDATA)0) && (vm->directByteBufferMemoryMax > (UDATA)I_64_MAX)) {
				parseErrorOption = VMOPT_XXMAXDIRECTMEMORYSIZEEQUALS;
				parseError = OPTION_OUTOFRANGE;
				goto _memParseError;
			}
#endif /* UINTPTR_MAX > INT64_MAX */
```

pshipton commented:
Was there a problem with the earlier iteration to use `(~(UDATA)0) >> 1`. 

I'll defer to @keithc-ca for other ideas.

keithc-ca commented on runtime/oti/vmargs_api.h:
```diff
@@ -89,13 +89,14 @@ getOptionString(J9VMInitArgs* j9vm_args, IDATA index);
 	GET_COMPOUND_OPTS returns all the option values from the command-line for a given option, separated by \0s. Based on certain rules.
 	GET_MAPPED_OPTION returns the value of a mapped option. Eg. if -Xsov1 maps to Xj91 then "-Xsov1:foo" returns "foo"
-	GET_MEM_VALUE returns a rounded value of a memory option. Eg. "-Xfoo32k" returns (32 * 1024).
+	GET_RAW_MEM_VALUE returns the value of a memory option. Eg. "-Xfoo32k" return (32 * 1024).
+	GET_MEM_VALUE returns a pointer size rounded value of a memory option. Eg. "-Xfoo32k" returns (32 * 1024) and "-Xfoo1023" returns 1024.
```
Please correct the spelling of "e.g." (at least for the part being changed).

This should provide an example showing how `GET_RAW_MEM_VALUE` differs from `GET_MEM_VALUE`.

keithc-ca commented on runtime/util/vmargs.c:
```diff
@@ -324,13 +324,14 @@ getCompoundOptions(J9PortLibrary *portLibrary, J9VMInitArgs* j9vm_args, IDATA el
 	GET_COMPOUND_OPTS returns all the option values from the command-line for a given option, separated by \0s. Based on certain rules.
 	GET_MAPPED_OPTION returns the value of a mapped option. Eg. if -Xsov1 maps to Xj91 then "-Xsov1:foo" returns "foo"
-	GET_MEM_VALUE returns a rounded value of a memory option. Eg. "-Xfoo32k" returns (32 * 1024).
+	GET_RAW_MEM_VALUE returns the value of a memory option. Eg. "-Xfoo32k" return (32 * 1024).
+	GET_MEM_VALUE returns a pointer size rounded value of a memory option. Eg. "-Xfoo32k" returns (32 * 1024) and "-Xfoo1023" returns 1024.
```
Same comments here: "e.g.", distinguishing example for `GET_RAW_MEM_VALUE`.

chrisdennis commented:
> Was there a problem with the earlier iteration to use `(~(UDATA)0) >> 1`?

That assumes that UDATA is 64-bit, right? What I'm pushing for is, if UDATA is capable of representing values higher than Java's Long.MAX_VALUE (implemented as a preprocessor condition), then check if the actual value is higher than can be stored in a Java long.

keithc-ca commented:
Do we really need this at all? What if the rounding near the upper limit were instead truncated?
Rather than round 9223372036854775807 up to 9223372036854775808 (which doesn't fit in a signed 64-bit location), we could round down to 9223372036854774784 (the largest multiple of 1024 that fits). I think it will be a very long while before we'll encounter systems where a request for 8388607TB could be satisfied.

keithc-ca commented:
I was proposing that truncation would only apply at the limit. The vast majority of use-cases would be unaffected.

pshipton commented:
It's fine with me. We can try it out and go with that if it doesn't cause any problems.

chrisdennis commented:
Right now I have:
```diff
diff --git a/runtime/util/vmargs.c b/runtime/util/vmargs.c
index 320fbc05f4..82c73490ea 100644
--- a/runtime/util/vmargs.c
+++ b/runtime/util/vmargs.c
@@ -578,8 +578,17 @@ optionValueOperations(J9PortLibrary *portLibrary, J9VMInitArgs* j9vm_args, IDATA
 					switch (*cursor) {
 						case '\0':
 							oldValue = value;
-							value = (value +  sizeof(UDATA) - 1) & ~(sizeof(UDATA) - 1);		/* round to nearest pointer value */
-							if (value < oldValue) return OPTION_OVERFLOW;
+							/* Round up to the next multiple of the pointer width */
+							value = (value + sizeof(UDATA) - 1) & ~(sizeof(UDATA) - 1);
+
+							if ((value < oldValue) || (((I_64) value) < ((I_64) oldValue))) {
+								/* 
+								 * Truncate (round in a negative direction) when overflow is detected
+								 * in either the signed 64-bit or default unsigned representations.
+								 * This avoids sign issues regardless of downstream usage.
+								 */
+								value = oldValue & ~(sizeof(UDATA) - 1);
+							}
 							break;
 						case 'k':
 						case 'K':
diff --git a/runtime/vm/jvminit.c b/runtime/vm/jvminit.c
index cbb388500e..a8c803153e 100644
--- a/runtime/vm/jvminit.c
+++ b/runtime/vm/jvminit.c
@@ -2555,6 +2555,10 @@ VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved)
 			if (OPTION_OK != parseError) {
 				parseErrorOption = VMOPT_XXMAXDIRECTMEMORYSIZEEQUALS;
 				goto _memParseError;
+			} else if (vm->directByteBufferMemoryMax != ~(UDATA)0 && vm->directByteBufferMemoryMax > I_64_MAX) {
+				parseErrorOption = VMOPT_XXMAXDIRECTMEMORYSIZEEQUALS;
+				parseError = OPTION_OUTOFRANGE;
+				goto _memParseError;
 			}
 
 			/* workaround option in case if OMRPORT_VMEM_ALLOC_QUICK Smart Address feature still be not reliable  */
```
This leads to the following behaviors:
```
$ java MaxDirectMemorySize
jdk.internal.misc.VM.maxDirectMemory() = 9663676416
$ java -XX:MaxDirectMemorySize=1 MaxDirectMemorySize
jdk.internal.misc.VM.maxDirectMemory() = 8
$ java -XX:MaxDirectMemorySize=9 MaxDirectMemorySize
jdk.internal.misc.VM.maxDirectMemory() = 16
$ java -XX:MaxDirectMemorySize=9223372036854775807 MaxDirectMemorySize
jdk.internal.misc.VM.maxDirectMemory() = 9223372036854775800
$ java -XX:MaxDirectMemorySize=9223372036854775808 MaxDirectMemorySize
JVMJ9VM013W Initialization error in function VMInitStages(0): Parse error for -XX:MaxDirectMemorySize= - value out of range.
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```
Let me know if this looks reasonable, and if so, should I create a new PR or just reuse this one?

pshipton commented:
Looks reasonable to me. You can continue to use this PR.

keithc-ca commented:
I'm not sure using `I_64_MAX` is appropriate; instead I suggest we should use `IDATA_MAX` which yields the same value on 64-bit platforms and the right 32-bit limit on 32-bit platforms.

keithc-ca commented:
If we're rounding by `sizeof(void *)`, then the limit on 32-bit platforms could be `U32_MAX - 4` and `I64_MAX - 8` for 64-bit platforms. All legal values can be represented in the Java `long` type.

keithc-ca commented on runtime/util/vmargs.c:
```diff
@@ -578,8 +578,17 @@ optionValueOperations(J9PortLibrary *portLibrary, J9VMInitArgs* j9vm_args, IDATA
 						case '\0':
 							oldValue = value;
-							value = (value +  sizeof(UDATA) - 1) & ~(sizeof(UDATA) - 1);		/* round to nearest pointer value */
-							if (value < oldValue) return OPTION_OVERFLOW;
+							/* Round up to the next multiple of the pointer width */
+							value = (value + sizeof(UDATA) - 1) & ~(sizeof(UDATA) - 1);
+
+							if ((value < oldValue) || (((I_64) value) < ((I_64) oldValue))) {
+								/*
+								 * Truncate (round in a negative direction) when overflow is detected
+								 * in either the signed 64-bit or default unsigned representations.
+								 * This avoids sign issues regardless of downstream usage.
+								 */
+								value = oldValue & ~(sizeof(UDATA) - 1);
+							}
 							break;
```
I don't think this has the desired behavior for 32-bit platforms.
```c
#if defined(J9VM_ENV_DATA64)
#define MEM_MAX_LIMIT I_64_MAX
#else /* defined(J9VM_ENV_DATA64) */
#define MEM_MAX_LIMIT U_32_MAX
#endif /* defined(J9VM_ENV_DATA64) */
							/* Round (up if possible) to a multiple of the pointer width. */
							if (value < ((UDATA)MEM_MAX_LIMIT & ~(sizeof(UDATA) - 1))) {
								value += sizeof(UDATA) - 1;
							}
							value &= ~(sizeof(UDATA) - 1);
#undef MEM_MAX_LIMIT
							break;
```

keithc-ca replied:
It applies a limit of 2^32-1 on 32-bit platforms and 2^63-1 on 64-bit platforms.
I would argue it's also easier to follow.

keithc-ca replied:
Also, on 32-bit platforms, `UDATA` and `I_64` differ both in width and signedness. A cast from `UDATA` to `I_64` is not well defined: Does the width change first and then get sign-extended or the other way around?

keithc-ca commented on runtime/vm/jvminit.c:
```diff
@@ -2555,6 +2555,10 @@ VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved)
 			if (OPTION_OK != parseError) {
 				parseErrorOption = VMOPT_XXMAXDIRECTMEMORYSIZEEQUALS;
 				goto _memParseError;
+			} else if (vm->directByteBufferMemoryMax != ~(UDATA)0 && vm->directByteBufferMemoryMax > I_64_MAX) {
+				parseErrorOption = VMOPT_XXMAXDIRECTMEMORYSIZEEQUALS;
+				parseError = OPTION_OUTOFRANGE;
+				goto _memParseError;
 			}
```
No need for `else` here. Similar to above, I think we want a different limit for 32-bit platforms.
```c
#if defined(J9VM_ENV_DATA64)
#define DBB_MAX_LIMIT I_64_MAX
#else /* defined(J9VM_ENV_DATA64) */
#define DBB_MAX_LIMIT U_32_MAX
#endif /* defined(J9VM_ENV_DATA64) */
			if ((~(UDATA)0 != vm->directByteBufferMemoryMax)
				&& (vm->directByteBufferMemoryMax > (((UDATA)DBB_MAX_LIMIT) & ~(sizeof(UDATA) - 1)))
			) {
				parseErrorOption = VMOPT_XXMAXDIRECTMEMORYSIZEEQUALS;
				parseError = OPTION_OUTOFRANGE;
				goto _memParseError;
			}
#undef DBB_MAX_LIMIT
```

chrisdennis commented:
I've added test coverage for both new behaviors in what seemed like the most appropriate places. This became a little complex though.

For the rounding behavior I had to fix the test logic in jvminit.c to avoid LLVM/Clang warnings on my platform, then I had to fix the test assertions to correctly account for the running platform. This makes me suspect that these tests are not currently being run? Not sure if there is anything I can do to fix that.

For the MaxDirectMemorySize bounds check I added a test, but had issues with it executing it locally due to the mode set in the `<variations>` tag of `playlist.xml`. This change can probably be dropped, but it wasn't clear to me why it was there in the first place.

For now I've left all these changes as separate commits for ease of reviewing, but they can all be squashed in, based on your preferences.

pshipton commented:
Correct that the tests in jvminit.c are not actively run, they need to be manually activated for testing when the code is modified.

pshipton commented:
jenkins test sanity.functional,extended.functional amac jdk11

keithc-ca reviewed (changes requested):
Please correct the spelling of "coverage" in the commit message for 0a2582dd2e1a3fbee27d688fbc29961bf42eb22a and leave a blank line between the summary (first) line and the remaining lines.

chrisdennis commented:
> Please correct the spelling of "coverage" in the commit message for [0a2582d](https://github.com/eclipse-openj9/openj9/commit/0a2582dd2e1a3fbee27d688fbc29961bf42eb22a) and leave a blank line between the summary (first) line and the remaining lines.

Fixed.

JasonFengJ9 reviewed (approved):
LGTM

chrisdennis commented:
@pshipton https://github.com/eclipse-openj9/openj9/pull/22555

pshipton commented:
Fix a compilation error https://github.com/eclipse-openj9/openj9/pull/22557
